src/grain/ledger.js

```diff
@@ -0,0 +1,111 @@
+  +amount: Grain,
+  +timestampMs: number,
+  +memo: string,
+|};
+
+// TODO: Import GrainAllocation once it merges.
+type GrainAllocation = {};
+
+/**
+ * Tracks a distrbution of newly-minted grain to contributors.
+ *
+ * Each Distribution contains zero or more allocations, which
+ * actually contain the mapping of new grain amounts to contributors.
+ */
+export type GrainDistributedV1 = {|
+  +type: "GRAIN_DISTRIBUTED",
+  +version: number,
+  +timestamMs: number,
+  +memo: string,
+  +allocations: $ReadOnlyArray<GrainAllocation>,
+|};
+
+/**
+ * Record that an "alias" identity is equivalent to a "canonical" identity.
+ *
+ * This event permanently records that one user NodeAddress (the alias) is
+ * equivalent to another NodeAddress (the canonical address). this is useful
+ * if, for example, a user has both a GitHub and a Discourse account. Once the
+ * identities are merged, they cannot be separated (as doing so risks creating
+ * negative balances).
+ *
+ * If a user has multiple aliases, then they should have multiple identity
+ * merge events.
+ *
+ * If an identity merge has the same address as alias and canonical, it is a
+ * no-op.
+ *
+ * It is allowed for the "canonical" address of one identity merge to later
+ * be the "alias" of another merge.
+ *
+ * After the identity merge, any events which reference an alias address will
+ * be treated as though they referenced the canonical address. (E.g., a
+ * transfer to an alias will be received by the canonical address.) This rule
+ * even applies to future identity merge events. As such, if we had an identity
+ * merge saying that "A" is an alias of "B", and a future merge saying that "B"
+ * is an alias of "A", this would be re-written to saying that "B" is an alias
+ * of "B", which is a no-op.
+ */
+export type IdentityMergedV1 = {|
+  +type: "IDENTITY_MERGED",
+  +version: number,
+  +timestampMs: number,
+  +alias: NodeAddressT,
+  +canonical: NodeAddressT,
+  +memo: string,
+|};
+
+export type LedgerEvent =
+  | GrainDistributedV1
+  | IdentityMergedV1
+  | GrainTransferredV1;
+
+/**
+ * A ledger tracks balances, earnings, and the event history of participants.
+ *
+ * We may refactor the `events` to be an async generator in the future,
+ * when we stop storing all the events in memory.
+ */
+export interface Ledger {
+  /**
+   * Stores the current grain balance of each address.
+   */
+  balances(): Map<NodeAddressT, Grain>;
+
+  /**
+   * Stores the lifetime earnings of each address.
+   *
+   * Necessary for computing which participants have been underpaid.
+   */
+  earnings(): Map<NodeAddressT, Grain>;
+
+  /**
+   * Retrieve the full history of all LedgerEvents.
+   */
+  events(): $ReadOnlyArray<LedgerEvent>;
+
+  /**
+   * For each address, list all known aliases.
+   *
+   * Addresses that are present in the history but have no aliases will be
+   * included, and will map to an empty array.
+   */
+  aliases(): Map<NodeAddressT, $ReadOnlyArray<NodeAddressT>>;
```

Comment:
How will this interplay with the identities field in project.json? Currently we are free to just edit that field as we need, but if the Ledger needs to know about when changes are made and prevent un-merging of identities etc, will we still be able to do it that way?

Reply:
I'm thinking once we put together the instance system, an "instance update" system will look like
- run thru the ledger to get the expected identity mapping
- compare to the project.json file
- if new identity contractions, write merge events to the ledger. if identity separations, error.
- compute cred for project, write graph and scores to disk.
- compute grain distribution based on that cred, write distribution events to ledger.

Reply:
Because projections are transient, I think it's worth taking a look at what use-case we have for this first.
We can disregard data we don't need, and tweak the data structure to be specific for our use-case.

For example if we do `(old) => new` resolving most often, changing the map from `Map<new, old[]>` to `Map<old, new>` would give more efficient lookups. Maybe both directions are common and we want to have both maps.

My guess is, `Map<old, new>` will be used for the projection logic itself, if we want to support the post-merge event redirecting of old addresses to the new. Also think that future producers of events (commands) should prefer using the new address, so they would use this lookup too. Not seeing a huge need for the inverse lookup.

Reply:
> if new identity contractions, write merge events to the ledger. if identity separations, error.

I do think this is a good approach, but do want to point out, it's not the ledger that should care about this. The ledger with the current business rules (post-merge the old addresses are considered aliases of the new address), would support this scenario just fine.

Suppose: `github/Beanow` and `discourse/Beanow` merged into `sourcecred/Beanow`. The 2 merge events are added to the ledger, the old addresses now redirect to `sourcecred/Beanow`.

Should the project.json remove this mapping, I would receive Cred like this:
- `github/Beanow: 321`
- `discourse/Beanow: 123`
- `sourcecred/Beanow: undefined`

And the grain distribution strategies would also allocate on `github/Beanow` and `discourse/Beanow` accordingly. The mapping would then redirect these allocations to `sourcecred/Beanow`.

Nothing crashes here, and in terms of events this looks like a totally valid history!

There's 2 problems with this though.
For one the Cred scores would not be accurate here. Changing the graph topology makes a difference for Cred flow so _all the Cred scores_, not just mine, would be skewed. Skewing the Grain allocations along with it.

Second is, there's a semantics problem, if the "unlinking" in the project.json was intentional and supposed to mean that these are not to be considered the same identity any more, then the end result of it going to a single account is unexpected.